src/wolff_producer.erl

```diff
@@ -501,73 +554,143 @@ get_produce_version(#{conn := Conn} = St) when is_pid(Conn) ->
-    inc_sent_success(Config, length(Calls), AttemptedBefore),
-    NewSentReqsCount = SentReqsCount - 1,
-    NewInflightCalls = InflightCalls - length(Calls),
-    wolff_metrics:inflight_set(Config, NewInflightCalls),
-    ok = replayq:ack(Q, Q_AckRef),
-    NewPendingAcks = evaluate_pending_ack_funs(PendingAcks, Calls, BaseOffset),
-    St#{sent_reqs := NewSentReqs,
-        sent_reqs_count := NewSentReqsCount,
-        inflight_calls := NewInflightCalls,
-        pending_acks := NewPendingAcks
-       }.
+note(Fmt, Args) ->
+  lists:flatten(io_lib:format(Fmt, Args)).
+
+-spec do_handle_kafka_ack(_, _, sent(), state()) -> state().
+do_handle_kafka_ack(?no_error, BaseOffset, #{q_items := Items}, St) ->
+  %% in case this item is the last one in a split-batch
+  %% the real base offset of the whole batch is shifted by the original batch size
+  OffsetShift =
+    case Items of
+      [?Q_ITEM(_CallId, _Ts, [#{offset_shift := Shift}])] -> Shift;
+      _ -> 0
+    end,
+  clear_sent_and_ack_callers(?no_error, BaseOffset - OffsetShift, St);
+do_handle_kafka_ack(?message_too_large = EC, _BaseOffset, Sent, St) ->
+  #{topic := Topic, partition := Partition, config := Config} = St,
+  #{q_items := Items} = Sent,
+  MsgCnt = count_msgs(Items),
+  case MsgCnt =:= 1 of
+    true ->
+      #kpro_req{msg = IoData} = make_request(Items, St),
+      Bytes = iolist_size(IoData),
+      %% This is a single message batch, but it's still too large
+      Note = note("A message (encoded_bytes=~p) is dropped because it's too large for this topic! "
+                  "Consider increase 'max.message.bytes' config on the server side!", [Bytes]),
+      log_error(Topic, Partition, EC, #{note => Note}),
+      clear_sent_and_ack_callers(EC, EC, St);
+    false ->
+      %% This is a batch of more than one messages
+      %% Split the batch, and re-send
+      #{max_batch_bytes := Max} = Config,
+      NewMax = max(1, (Max + 1) div 2),
+      St1 = St#{config := Config#{max_batch_bytes := NewMax}},
+      Note = note("Config max_batch_bytes=~p is too large for this topic, "
+                  "trying to split the current batch (of ~p messages) "
+                  "into single-message batches and retry. "
+                  "Will use max_batch_bytes=~p to collect future batches.",
+                  [Max, MsgCnt, NewMax]),
+      log_warn(Topic, Partition, EC, #{note => Note}),
+      resend_sent_reqs(St1, EC)
+  end;
+do_handle_kafka_ack(ErrorCode, _BaseOffset, Sent, St) ->
+  %% Other errors, such as not_leader_for_partition
+  #{attempts := Attempts, q_items := Items} = Sent,
+  #{topic := Topic,
+    partition := Partition,
+    config := Config
+   } = St,
+  NrOfCalls = count_calls(Items),
+  inc_sent_failed(Config, NrOfCalls, Attempts),
+  log_warn(Topic, Partition, "error_in_produce_response",
+           #{error_code => ErrorCode,
+             batch_size => count_msgs(Items),
+             attempts => Attempts}),
+  erlang:throw(ErrorCode).
```

Comment:
Don't we need to reply callers before throwing?

Reply:
this throw will cause retry, reply caller will be done after retry.